File: rs/moq-audio/src/playout/fixture.rs
//! Signals and measurements the playout tests share.
//!
//! Everything here is deterministic: the noise source is a fixed LCG, so a failure
//! reproduces exactly.

use std::f32::consts::TAU;

/// Interleaved samples of a sine, every channel in phase.
pub(crate) fn tone(rate: u32, seconds: f32, frequency: f32, amplitude: f32, channels: usize) -> Vec<f32> {
	let total = (rate as f32 * seconds) as usize;
	let step = TAU * frequency / rate as f32;
	(0..total)
		.flat_map(|i| {
			let sample = amplitude * (step * i as f32).sin();
			std::iter::repeat_n(sample, channels)
		})
		.collect()
}

/// Interleaved stereo where the right channel is the left one inverted.
pub(crate) fn antiphase(rate: u32, seconds: f32, frequency: f32, amplitude: f32) -> Vec<f32> {
	let mono = tone(rate, seconds, frequency, amplitude, 1);
	mono.into_iter().flat_map(|s| [s, -s]).collect()
}

/// Interleaved white noise in `-amplitude..amplitude`, the same in every channel.
pub(crate) fn noise(rate: u32, seconds: f32, amplitude: f32, channels: usize) -> Vec<f32> {
	let total = (rate as f32 * seconds) as usize;
	let mut state = 0x1234_5678_9abc_def0u64;
	(0..total)
		.flat_map(|_| {
			state = state
				.wrapping_mul(6364136223846793005)
				.wrapping_add(1442695040888963407);
			let uniform = ((state >> 40) as f32) / (1u64 << 23) as f32 - 1.0;
			std::iter::repeat_n(uniform * amplitude, channels)
		})
		.collect()
}

/// A signal with a short period plus enough noise to spoil its correlation.
pub(crate) fn buzz(rate: u32, seconds: f32, period: std::time::Duration, amplitude: f32, jitter: f32) -> Vec<f32> {
	let total = (rate as f32 * seconds) as usize;
	let len = super::frames(rate, period);
	let shape = noise(rate, period.as_secs_f32(), amplitude, 1);
	let dither = noise(rate, seconds, jitter, 1);
	(0..total).map(|i| shape[i % len] + dither[i]).collect()
}

/// Mean per sample energy.
pub(crate) fn energy(pcm: &[f32]) -> f32 {
	pcm.iter().map(|x| x * x).sum::<f32>() / pcm.len().max(1) as f32
}

/// Goertzel magnitude of `frequency` in `pcm`, normalised by length.
pub(crate) fn goertzel(pcm: &[f32], rate: u32, frequency: f32) -> f32 {
	let coefficient = 2.0 * (TAU * frequency / rate as f32).cos();
	let (mut first, mut second) = (0.0f32, 0.0f32);
	for sample in pcm {
		let next = sample + coefficient * first - second;
		second = first;
		first = next;
	}
	let power = first * first + second * second - coefficient * first * second;
	power.max(0.0).sqrt() / pcm.len() as f32
}

/// The strongest frequency in `low..high`, to 0.5 Hz.
pub(crate) fn dominant(pcm: &[f32], rate: u32, low: f32, high: f32) -> f32 {
	let mut best = (low, f32::MIN);
	let mut frequency = low;
	while frequency <= high {
		let magnitude = goertzel(pcm, rate, frequency);
		if magnitude > best.1 {
			best = (frequency, magnitude);
		}
		frequency += 0.5;
	}
	best.0
}

/// The largest jump between consecutive samples.
pub(crate) fn max_step(pcm: &[f32]) -> f32 {
	pcm.windows(2).map(|w| (w[1] - w[0]).abs()).fold(0.0f32, f32::max)
}

/// The longest run of samples that are exactly zero.
pub(crate) fn zero_run(pcm: &[f32]) -> usize {
	let mut longest = 0;
	let mut run = 0;
	for sample in pcm {
		run = if *sample == 0.0 { run + 1 } else { 0 };
		longest = longest.max(run);
	}
	longest
}

File: js/watch/src/audio/playout/fixture.ts
/**
 * Signals and measurements the playout tests share, mirroring `rs/moq-audio/src/playout/fixture.rs`
 * so a failure in one language reproduces in the other.
 *
 * Everything here is deterministic: the noise source is a fixed LCG, so a failure reproduces exactly.
 *
 * @internal Test support, not part of the engine.
 */

const TAU = Math.PI * 2;

/** Planar samples of a sine, every channel in phase. */
export function tone(
	rate: number,
	seconds: number,
	frequency: number,
	amplitude: number,
	channels = 1,
): Float32Array[] {
	const total = Math.floor(rate * seconds);
	const step = (TAU * frequency) / rate;
	const mono = new Float32Array(total);
	for (let i = 0; i < total; i++) mono[i] = amplitude * Math.sin(step * i);
	return Array.from({ length: channels }, () => mono.slice());
}

/** Planar stereo where the right channel is the left one inverted. */
export function antiphase(rate: number, seconds: number, frequency: number, amplitude: number): Float32Array[] {
	const [left] = tone(rate, seconds, frequency, amplitude, 1);
	const right = new Float32Array(left.length);
	for (let i = 0; i < left.length; i++) right[i] = -left[i];
	return [left, right];
}

/** Planar white noise in `-amplitude..amplitude`, the same in every channel. */
export function noise(rate: number, seconds: number, amplitude: number, channels = 1): Float32Array[] {
	const total = Math.floor(rate * seconds);
	const mono = new Float32Array(total);
	let state = 0x1234_5678_9abc_def0n;
	for (let i = 0; i < total; i++) {
		state = (state * 6364136223846793005n + 1442695040888963407n) & 0xffff_ffff_ffff_ffffn;
		const uniform = Number(state >> 40n) / 2 ** 23 - 1;
		mono[i] = uniform * amplitude;
	}
	return Array.from({ length: channels }, () => mono.slice());
}

/**
 * A tone that pauses, over a noise floor: the shape a background estimate can learn from.
 *
 * A pure sine never trains {@link Noise}, because no window of it is ever quiet, so concealment
 * built on one fades into digital silence rather than into the room. Real speech pauses, and the
 * pauses are where the room is measured, so anything testing concealment over more than a few
 * blocks wants this rather than {@link tone}.
 */
export function speech(
	rate: number,
	seconds: number,
	frequency: number,
	amplitude: number,
	floor: number,
	channels = 1,
): Float32Array[] {
	const [voiced] = tone(rate, seconds, frequency, amplitude, 1);
	const [room] = noise(rate, seconds, floor, 1);
	const cycle = Math.round(rate * 1.0);
	const pause = Math.round(rate * 0.2);

	const mono = new Float32Array(voiced.length);
	for (let i = 0; i < mono.length; i++) {
		mono[i] = room[i] + (i % cycle < cycle - pause ? voiced[i] : 0);
	}
	return Array.from({ length: channels }, () => mono.slice());
}

/** A signal with a short period plus enough noise to spoil its correlation. */
export function buzz(
	rate: number,
	seconds: number,
	periodMs: number,
	amplitude: number,
	jitter: number,
): Float32Array[] {
	const total = Math.floor(rate * seconds);
	const [shape] = noise(rate, periodMs / 1000, amplitude, 1);
	const [dither] = noise(rate, seconds, jitter, 1);
	const out = new Float32Array(total);
	for (let i = 0; i < total; i++) out[i] = shape[i % shape.length] + dither[i];
	return [out];
}

/** Mean per sample energy of the first `length` frames. */
export function energy(pcm: Float32Array, length = pcm.length): number {
	let sum = 0;
	for (let i = 0; i < length; i++) sum += pcm[i] * pcm[i];
	return sum / Math.max(1, length);
}

/** Goertzel magnitude of `frequency` in `pcm`, normalised by length. */
export function goertzel(pcm: Float32Array, rate: number, frequency: number): number {
	const coefficient = 2 * Math.cos((TAU * frequency) / rate);
	let first = 0;
	let second = 0;
	for (const sample of pcm) {
		const next = sample + coefficient * first - second;
		second = first;
		first = next;
	}
	const power = first * first + second * second - coefficient * first * second;
	return Math.sqrt(Math.max(0, power)) / pcm.length;
}

/** The strongest frequency in `low..high`, to 0.5Hz. */
export function dominant(pcm: Float32Array, rate: number, low: number, high: number): number {
	let best = low;
	let magnitude = Number.NEGATIVE_INFINITY;
	for (let frequency = low; frequency <= high; frequency += 0.5) {
		const found = goertzel(pcm, rate, frequency);
		if (found > magnitude) {
			magnitude = found;
			best = frequency;
		}
	}
	return best;
}

/** The largest jump between consecutive samples. */
export function maxStep(pcm: Float32Array): number {
	let largest = 0;
	for (let i = 1; i < pcm.length; i++) largest = Math.max(largest, Math.abs(pcm[i] - pcm[i - 1]));
	return largest;
}

/** The longest run of samples that are exactly zero. */
export function zeroRun(pcm: Float32Array): number {
	let longest = 0;
	let run = 0;
	for (const sample of pcm) {
		run = sample === 0 ? run + 1 : 0;
		longest = Math.max(longest, run);
	}
	return longest;
}
